Framework/HistogramData/inc/MantidHistogramData/HistogramIterator.h

```diff
@@ -0,0 +1,72 @@
+
+#ifndef MANTID_HISTOGRAMDATA_HISTOGRAMITERATOR_H_
+#define MANTID_HISTOGRAMDATA_HISTOGRAMITERATOR_H_
+
+#include "MantidHistogramData/DllConfig.h"
+#include "MantidHistogramData/HistogramItem.h"
+#include "MantidKernel/make_unique.h"
+
+#include <boost/iterator/iterator_facade.hpp>
+#include <memory>
+
+namespace Mantid {
+namespace HistogramData {
+
+/** HistogramIterator
+
+  HistogramIterator implements an the iterator interface for HistogramData.
+  At each position the iterator will point to an instance of a HistogramItem.
+  This item provides direct access to the values at a particular index.
+
+  @author Samuel Jackson
+  @date 2017
+
+  Copyright &copy; 2016 ISIS Rutherford Appleton Laboratory, NScD Oak Ridge
+  National Laboratory & European Spallation Source
+
+  This file is part of Mantid.
+
+  Mantid is free software; you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation; either version 3 of the License, or
+  (at your option) any later version.
+
+  Mantid is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+  File change history is stored at: <https://github.com/mantidproject/mantid>
+  Code Documentation is available at: <http://doxygen.mantidproject.org>
+*/
+class MANTID_HISTOGRAMDATA_DLL HistogramIterator
+    : public boost::iterator_facade<HistogramIterator, HistogramItem,
+                                    boost::bidirectional_traversal_tag> {
+
+public:
+  HistogramIterator(const Histogram &histogram, const size_t index = 0)
+      : m_histogram(histogram), m_index(index),
+        m_currentItem(Kernel::make_unique<HistogramItem>(histogram, index)){};
```

Comment:
Why the `unique_ptr`? Can we avoid the memory allocation?

Reply:
I want to avoid this but I'm not sure how. I can't simply create and return a reference to `HistogramItem` as that would mean returning reference to a stack variable. `HistogramItem` is not copyable so I can't return a copy. I also can't have an instance variable `HistogramItem` because the `HistogramIterator` is copied when returned from the `begin()` call.

Reply:
But the pointer is not solving that, is it? You will have an invalided reference to `HistogramItem` once you, e.g., increment the iterator.

Reply:
Why not just store `HistogramItem` in `HistogramIterator` (instead of `m_histogram` and `m_index`) and return a reference to that?

Reply:
Ok, but what about the last point? `begin` constructs and returns a `HistogramIterator`. If it has an instance of `HistogramItem` then `HistogramIterator` has an implicitly deleted copy constructor?

Reply:
Right, I guess then `HistogramItem` should be copyable after all?

Reply:
I guess. 😄